src/file_pipe_log/pipe.rs

```diff
@@ -272,31 +371,73 @@ impl<F: FileSystem> SinglePipe<F> {
 
     fn file_span(&self) -> (FileSeq, FileSeq) {
         let files = self.files.read();
-        (files.first_seq, files.active_seq)
+        (
+            files.first_seq_in_use,
+            files.first_seq + files.fds.len() as u64 - 1,
+        )
     }
 
     fn total_size(&self) -> usize {
         let files = self.files.read();
-        (files.active_seq - files.first_seq + 1) as usize * self.target_file_size
+        files.fds.len() * self.target_file_size
     }
 
     fn rotate(&self) -> Result<()> {
         self.rotate_imp(&mut self.active_file.lock())
     }
 
     fn purge_to(&self, file_seq: FileSeq) -> Result<usize> {
-        let (purged, remained) = {
+        let (
+            first_purge_seq, /* first seq for purging */
+            purged,          /* count of purged files */
+            recycled,        /* count of recycled files */
+            remained,        /* count of remained files */
+        ) = {
             let mut files = self.files.write();
-            if file_seq > files.active_seq {
+            if file_seq >= files.first_seq + files.fds.len() as u64 {
                 return Err(box_err!("Purge active or newer files"));
             }
-            let end_offset = file_seq.saturating_sub(files.first_seq) as usize;
-            files.fds.drain(..end_offset);
-            files.first_seq = file_seq;
-            (end_offset, files.fds.len())
+            let mut first_purge_seq = 0;
```

Comment:
Nope, if the `recycle capacity` was enought to incoming stale files, the `first_purge_seq` should `== 0`, because there has no stale files should be purged.

Reply:
In that case, `purged` equals zero, `first_purge_seq` wouldn't be used at all.